docs/rfcs/0011-script-run-stage.md

```diff
@@ -0,0 +1,208 @@
+```
+
+## prepare environment for execution
+
+Commands are executed on the container of piped or on the host OS(standalone).
+
+## CUSTOM_SYNC in the future
+
+"CUSTOM_SYNC" stage will be deprecated because the "SCRIPT_RUN" has also similar features.
+I expect users to use "SCRIPT_RUN" for executing any command before or after other stages.
+
+
+# Alternatives
+
+## What's the difference between "CUSTOM_SYNC" and "SCRIPT_SYNC"?
+
+"CUSTOM_SYNC" is one of the stages to **sync**, but "SCRIPT_RUN" is the stage to **execute commands**.
+
+## How about other CD tools?
+
+### Argo
+
+#### strategy
+
+- **execute command with another k8s resources**
+#### details
+Resource Hooks 
+https://argo-cd.readthedocs.io/en/stable/user-guide/resource_hooks/#resource-hooks
+> Hooks are ways to run scripts before, during, and after a Sync operation. Hooks can also be run if a Sync operation fails at any point.
+
+- There are four points to execute command.
+	- PreSync: before sync
+	- Sync: during sync
+	- PostSync: after sync
+	- SyncFaill: failed to sync
+- **To execute command, ArgoCD applys k8s resources such as Job or Pod, [[Argo Workflows]] ...**
+- users set some annotations and ArgoCD detects them to control the order to execute command
+	- e.g. https://argo-cd.readthedocs.io/en/stable/user-guide/resource_hooks/#using-a-hook-to-send-a-slack-message
+
+#### pros/cons
+
+**pros**
+- can separate respolibility for delivery and executing any command
+
+**cons**
+- users need to prepare and manage the resource to execute any command
+
+### FluxCD
+- There is no functions to realize that
+
+### Flagger
+
+#### strategy
+- Call api set as webhooks on each points and execute command on the api
+- Flagger just call api registerd as webhooks.
+
+#### details
+
+**Webhooks**
+https://fluxcd.io/flagger/usage/webhooks/#load-testing
+>Flagger will call each webhook URL and determine from the response status code (HTTP 2xx) if the canary is failing or not.
+
+- There are some webhook points.
+	- confirm-rollout
+	- pre-rollout
+	- rollout
+	- confirm-traffic-increase
+	- confirm-promotion
+	- post-rollout
+	- rollback
+	- event
+- Flagger calls webhooks on each points.
+	- e.g. load testing with Flagger: https://fluxcd.io/flagger/usage/webhooks/#load-testing
+- if users want to execute command with webhook, set the command text to `metadata` section and prepare webhook handler to execute cmd parsed from metadata.
+
+#### pros/cons
+
+**pros**
+- can separate respolibility for delivery and executing any command
+
+**cons**
+- users need to prepare api for the webhooks.
+
+
+# Unresolved questions
+
+- It might be better to change from alpine to debian and so on to provide users popular commands (e.g. curl) by default.
+- It might be better to add `runsOnRollback` on each sync stage if users want to control when rollback.
+This is a just idea.
```

Comment:
I don't get this part, do you mean in case of multiple script_run stages are combined, we should add `runsOnRollback` to each? And is this `runsOnRollback` is same as `onRollback` (the one used in the example above)?

Reply:
@khanhtc1202 
This means It might be a way to add `onRollback` to each stage, not to create a new stage.

e.g. K8S_CANARY_ROLLOUT
```
  pipeline:
    stages:
      - name: K8S_CANARY_ROLLOUT
        with:
          replicas: 10%
          onRollback: "curl -X POST -H 'Content-type: application/json' --data '{"text":"failed to deploy: rollback"}' $SLACK_WEBHOOK_URL"
```

Reply:
Why does it come to this place 🤔 And also, how do we combine your idea with the current rollback logic? Do we have any plans for this?

Reply:
> Why does it come to this place 🤔 And also, how do we combine your idea with the current rollback logic? Do we have any plans for this?

It is a just idea, so I have no idea to realize it 🙇 
But I thought it might be more useful to be able to set `onRollback` on each stages than set a stage to execute any command.